### PR TITLE
Flink: Integrate Flink reader to SQL

### PR DESCRIPTION
**Pull request by JingsongLi**
This is subtask of #1293
Also fix left comments in #1346

### CODE REVIEW

File: flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java [hunk 1]
```diff
@@ -100,8 +100,8 @@ public LogicalType primitive(Type.PrimitiveType primitive) {
       case DATE:
         return new DateType();
       case TIME:
-        // MICROS
-        return new TimeType(6);
+        // Flink only support TimeType with default precision now.
```
Comment (rdblue) [inline flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java:103]: Can you update the comment to state what Flink's default precision is?

File: flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java [hunk 2]
```diff
@@ -252,7 +252,7 @@ public void testInconsistentTypes() {
         Types.BinaryType.get(), new VarBinaryType(VarBinaryType.MAX_LENGTH),
         new VarBinaryType(100), Types.BinaryType.get());
     checkInconsistentType(
-        Types.TimeType.get(), new TimeType(6),
+        Types.TimeType.get(), new TimeType(),
```
Comment (rdblue) [inline flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java:255]: Are these related changes, or should the time changes go into a separate PR?

File: flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkScan.java [hunk 3]
```diff
@@ -230,7 +215,7 @@ public void testSnapshotReads() throws Exception {
     long timestampMillis = table.currentSnapshot().timestampMillis();
 
     // produce another timestamp
-    Thread.sleep(10);
+    waitUntilAfter(10);
```
Comment (rdblue) [inline flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkScan.java:218]: This should be called with `timestampMillis` because you want to wait until it is strictly after the current snapshot's timestamp.

**Comment by rdblue** [comment]
Mostly looks good. Just a couple of minor issues.

File: flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkScan.java (hunk 3 above)
Comment (JingsongLi) [reply flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkScan.java:218]: Yes, you are right.

File: flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java (hunk 2 above)
Comment (JingsongLi) [reply flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java:255]: I will create one.

**Comment by JingsongLi** [comment]
> Let's take a look at `ScanOptions` in the next PR then. I would prefer to keep user-facing APIs simple, rather than leaking a SQL concern (options come from `WITH`) to users (need to use two builders). Since SQL will most likely use the `fromProperties` method, it may make sense to use a single builder, add `withProperties`, and pass properties from SQL as a map.

Got it, I think I can change the `ScanOptions` to `ScanContext`, it is an internal helper class like `TableScanContext`.
Then expose all setters to `FlinkSource.Builder`.

File: flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java (hunk 2 above)
Comment (JingsongLi) [reply flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java:255]: Created #1521

Comment (rdblue) [reply flink/src/test/java/org/apache/iceberg/flink/TestFlinkSchemaUtil.java:255]: Merged #1521.

File: flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkScan.java [hunk 4]
```diff
@@ -230,13 +246,16 @@ public void testSnapshotReads() throws Exception {
     long timestampMillis = table.currentSnapshot().timestampMillis();
 
     // produce another timestamp
-    Thread.sleep(10);
+    waitUntilAfter(System.currentTimeMillis() + 10);
```
Comment (rdblue) [inline flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkScan.java:249]: Why is this waiting until 10ms from now? That's basically equivalent to `Thread.sleep`. The purpose of using `waitUntilAfter` is to ensure the next call to `System.currentTimeMillis()` will be after some point in time. In most cases, that's the last snapshot's timestamp.

I think this should be `waitUntilAfter(table.currentSnapshot().timestampMillis());`. Then the next append will be strictly after that snapshot's timestamp, and the query with `as-of-timestamp` will read the older snapshot. There should be no need to wait as long as 10ms.

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java [hunk 5]
```diff
@@ -89,8 +88,18 @@ public Builder table(Table newTable) {
       return this;
     }
 
-    public Builder filters(List<Expression> newFilters) {
-      this.filterExpressions = newFilters;
+    public Builder hadoopConf(Configuration newConf) {
```
Comment (rdblue) [inline flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java:91]: We try to avoid using Hadoop classes in Iceberg APIs because they are hard to remove. Injecting a Hadoop `Configuration` is currently done in one place: to instantiate a catalog that requires it.

The catalog creates tables and tables are associated with a `FileIO`, so the configuration is passed down through that chain. The table's configuration should be used for any table configuration.

MR also has a Hadoop `Configuration`, but that's required by the API so we can't remove it from the API there. But we still prefer using the table's `Configuration` when it is needed for components like `HadoopFileIO`.

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java [hunk 6]
```diff
@@ -45,30 +43,25 @@
...
   private transient RowDataIterator iterator;
 
-  FlinkInputFormat(
-      TableLoader tableLoader, Schema projectedSchema, FileIO io, EncryptionManager encryption,
-      List<Expression> filterExpressions, ScanOptions options, SerializableConfiguration serializableConf) {
+  FlinkInputFormat(TableLoader tableLoader, SerializableConfiguration serializableConf, FileIO io,
+                   EncryptionManager encryption, ScanContext context) {
     this.tableLoader = tableLoader;
-    this.projectedSchema = projectedSchema;
+    this.serializableConf = serializableConf;
```
Comment (rdblue) [inline flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:56]: Why pass the conf here? I don't see it used.

File: flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkScan.java (hunk 4 above)
Comment (JingsongLi) [reply flink/src/test/java/org/apache/iceberg/flink/source/TestFlinkScan.java:249]: I see. You are right.

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 6 above)
Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:56]: It is used to open `TableLoader`.

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java (hunk 5 above)
Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java:91]: +1 for avoiding using Hadoop `Configuration`.

Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java:91]: The reason why Hadoop conf needs to be passed now is because:
- JobManager needs the splits of the scan, so it needs to call `Table.newScan`.
- So JobManager needs `Table` object, where can a table be generated? `TableLoader` -> from `Catalog` or `HadoopTables`.
- The creation of `Catalog` and `HadoopTables` needs a Hadoop `Configuration`.

Maybe we can pass this chain with an Iceberg object (`FileIO` may looks not good to the creation of catalog)?

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java (hunk 6 above)
Comment (rdblue) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java:56]: I see now. Thanks!

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java (hunk 5 above)
Comment (rdblue) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java:91]: Maybe `CatalogLoader` should serialize its own `Configuration`? It would make sense to pass one when creating a `CatalogLoader` for `HadoopCatalog` or `HiveCatalog` because those need a `Configuration` to create the catalog.

Comment (rdblue) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java:91]: Or maybe this could use the approach from the `FlinkCatalogFactory`:

```java
  public static Configuration clusterHadoopConf() {
    return HadoopUtils.getHadoopConfiguration(GlobalConfiguration.loadConfiguration());
  }
```

Using `clusterHadoopConf` internally here would avoid the need to expose this in the API and we could add `Configuration` to the loader later.

**Comment by rdblue** [comment]
Looks good, @JingsongLi! I'd like to remove the Hadoop `Configuration` from the API, but that was an existing problem so there's no need to block this PR. Thank you!

File: flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java (hunk 5 above)
Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/source/FlinkSource.java:91]: Using `clusterHadoopConf` may not so flexible.
I am +1 for CatalogLoader should serialize its own Configuration. I will create a PR later for source and sink.
